activerecord/lib/active_record/autosave_association.rb

```diff
@@ -176,7 +176,8 @@ def define_non_cyclic_method(name, &block)
         # check if the save or validation methods have already been defined
         # before actually defining them.
         def add_autosave_association_callbacks(reflection)
-          save_method = :"autosave_associated_records_for_#{reflection.name}"
+          method_suffix = reflection.options[:class_name].try(:underscore)
```

Comment:
Try out this way: `['autosave_associated_records_for', reflection.name, method_suffix].compact.join('_')`.

or conditionally:

```ruby
if method_suffix.present?
  save_method = :"autosave_associated_records_for_#{reflection.name}_#{method_suffix}"
else
  save_method = :"autosave_associated_records_for_#{reflection.name}"
end
```